### PR TITLE
applications: nrf_desktop: Add led stream

### PR DESCRIPTION
**Pull request by mip1-nordic**
Add led stream option to configurator.

Jira: DESK-431

Signed-off-by: Michal Patyk <michal.patyk@nordicsemi.no>

### CODE REVIEW

**Comment by mip1-nordic** [comment]
@MarekPieta

File: scripts/hid_configurator/configurator_cli.py [hunk 1]
```diff
@@ -93,6 +103,9 @@ def parse_arguments():
         sp_commands.add_parser('fwinfo', help='Obtain information about FW image')
         sp_commands.add_parser('fwreboot', help='Request FW reboot')
 
+        if device_name == 'gaming_mouse':
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_cli.py]: Maybe we could add this option anyway and just later on check if device supports it ?

Comment (MarekPieta) [inline scripts/hid_configurator/configurator_cli.py]: Maybe we could add flag informing if this option is supported in board ocnfigurations ?

File: scripts/hid_configurator/configurator_core.py [hunk 2]
```diff
@@ -620,5 +623,74 @@ def is_dfu_image_correct(dfu_image):
     return True
 
 
+class Step:
+    def __init__(self, r, g, b, substep_count, substep_time):
+        self.r = int(r)
+        self.g = int(g)
+        self.b = int(b)
+        self.substep_count = int(substep_count)
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_core.py]: Why do we need conversion here ?

File: scripts/hid_configurator/configurator_core.py [hunk 3]
```diff
@@ -620,5 +623,74 @@ def is_dfu_image_correct(dfu_image):
...
+        success = exchange_feature_report(dev, recipient, event_id, event_data, False)
+    except:
+        success = False
+
+    return success
+
+
+def led_send_random_color_steps(dev, recipient, n):
+    logging.debug('Sending random color steps')
+    substep_count = 10
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_core.py]: Maybe we could define these default values e.g. in `__init__`?

File: scripts/hid_configurator/configurator_core.py [hunk 4]
```diff
@@ -620,5 +623,74 @@ def is_dfu_image_correct(dfu_image):
...
+
+    return struct.unpack(fmt, fetched_data)[0]
+
+def send_continuous_led_stream(dev, recipient):
+    while 1:
+        free_places = fetch_free_steps_buffor_info(dev, recipient)
+        if free_places == None:
+            return None
+        logging.debug('free places: {}'.format(free_places))
+        while free_places < 1:
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_core.py]: Maybe something like:
`if free_places == 0:
    time.sleep(0.1)
else:
    led_send_random_color_steps(...)`

Comment (MarekPieta) [reply scripts/hid_configurator/configurator_core.py]: btw. how are we supposed to finish this loop ?

File: scripts/hid_configurator/configurator_cli.py (hunk 1 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_cli.py]: done

File: scripts/hid_configurator/configurator_core.py (hunk 2 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: no, done

File: scripts/hid_configurator/configurator_core.py (hunk 3 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: step substep_count and substep_time will change dynamically, so I don't want them in __init__

File: scripts/hid_configurator/configurator_core.py (hunk 4 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: added signal_handler, done

**Comment by mip1-nordic** [comment]
@MarekPieta

File: scripts/hid_configurator/configurator_core.py [hunk 5]
```diff
@@ -625,11 +640,11 @@ def is_dfu_image_correct(dfu_image):
...
-        self.r = int(r)
-        self.g = int(g)
-        self.b = int(b)
-        self.substep_count = int(substep_count)
-        self.substep_time = int(substep_time)
+        self.r = r
+        self.g = g
+        self.b = b
+        self.substep_count = substep_count
+        self.substep_time = substep_time
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_core.py]: add default values

File: scripts/hid_configurator/configurator_core.py [hunk 6]
```diff
@@ -92,30 +93,44 @@
...
-        'config' : None
+        'config' : None,
+        'stream' : False,
     }
 }
 
+interrupted = False
+
+
+def signal_handler(signal, frame):
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_core.py]: move closer to while loop and avoid using globals if possible

Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: done

File: scripts/hid_configurator/configurator_core.py (hunk 5 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: done

**Comment by mip1-nordic** [comment]
@MarekPieta

File: scripts/hid_configurator/configurator_core.py [hunk 7]
```diff
@@ -677,19 +681,26 @@ def fetch_free_steps_buffor_info(dev, recipient):
...
+interrupted = False
+
+
 def send_continuous_led_stream(dev, recipient):
-    while 1:
+    def signal_handler(signal, frame):
+        global interrupted
+        interrupted = True
+
+    signal.signal(signal.SIGINT, signal_handler)
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_core.py]: later, we may think about moving this operation to separate thread, but currently we can leave it as it is

File: scripts/hid_configurator/configurator_core.py [hunk 8]
```diff
@@ -677,19 +681,26 @@ def fetch_free_steps_buffor_info(dev, recipient):
...
+
+    while not interrupted:
         free_places = fetch_free_steps_buffor_info(dev, recipient)
         if free_places == None:
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_core.py]: is None?

File: scripts/hid_configurator/configurator_core.py [hunk 9]
```diff
@@ -677,19 +681,26 @@ def fetch_free_steps_buffor_info(dev, recipient):
...
-            return None
+            continue
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_core.py]: Can we continue here, or we should return None ?

Comment (MarekPieta) [reply scripts/hid_configurator/configurator_core.py]: Maybe add sleep here at least

File: scripts/hid_configurator/configurator_core.py (hunk 8 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: done

File: scripts/hid_configurator/configurator_core.py (hunk 9 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: discussed offline

**Comment by MarekPieta** [comment]
I close this PR, I think we may create PR for main repository.

**Pull request closed without merging**